### PR TITLE
Remove requirement for a cluster label value

### LINKED ISSUE #1736: HA dedupe should not require a "cluster" label
I'm told the purpose of the cluster label is to allow differentiatiation between multiple Prometheus HA pairs in a tenant. Its name is set via `-distributor.ha-tracker.cluster`.

Cortex should support the scenario where there is only one HA pair in a tenant, by making the cluster label optional.

### PR DESCRIPTION
**Pull request by uepoch**
Fixes #1736

Not tested yet, I will add some

### CODE REVIEW

File: pkg/util/validation/limits.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 	f.IntVar(&l.IngestionBurstSize, "distributor.ingestion-burst-size", 50000, "Per-user allowed ingestion burst size (in number of samples). Warning, very high limits will be reset every -distributor.limiter-reload-period.")
 	f.BoolVar(&l.AcceptHASamples, "distributor.ha-tracker.enable-for-all-users", false, "Flag to enable, for all users, handling of samples with external labels identifying replicas in an HA Prometheus setup.")
 	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
-	f.StringVar(&l.HAClusterLabel, "distributor.ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Prometheus HA cluster.")
+	f.StringVar(&l.HAClusterLabel, "distributor.ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Prometheus HA cluster. If set to empty string. It will only dedup with replica label")
```
Comment (pracucci) [inline pkg/util/validation/limits.go]: `If set to empty string. It will` > `If set to empty string, it will`

File: pkg/distributor/ha_tracker.go [hunk 2]
```diff
@@ -242,18 +242,20 @@ func replicasNotMatchError(replica, elected string) error {
...
-// the replica or cluster label and returning their values. Returns an error
-// if we find one but not both of the labels.
+// Searches for the values of replicaLabel and clusterLabel
+// returning their values. Returns an error if we find one but not both of the labels.
 func findHALabels(replicaLabel, clusterLabel string, labels []client.LabelAdapter) (string, string) {
 	var cluster, replica string
 	var pair client.LabelAdapter
 
-	for _, pair = range labels {
+	for i := 0; i < len(labels) && replica == "" && cluster == ""; i++ {
```
Comment (pracucci) [inline pkg/distributor/ha_tracker.go]: I would suggest to cover the new case the `clusterLabel` is `""` in `TestFindHALabels` (`ha_tracker_test.go`)

File: pkg/distributor/ha_tracker.go [hunk 3]
```diff
@@ -242,18 +242,20 @@ func replicasNotMatchError(replica, elected string) error {
...
 		if pair.Name == replicaLabel {
 			replica = string(pair.Value)
+			if clusterLabel == "" {
```
Comment (pracucci) [inline pkg/distributor/ha_tracker.go]: Am I wrong, or this `if` can be removed checking it in the `for` statement? For example, in the `for` you may replace `&& cluster == ""` with `&& (cluster == "" || clusterLabel == "")`.

File: pkg/distributor/ha_tracker.go (hunk 2 above)
Comment (uepoch) [reply pkg/distributor/ha_tracker.go]: fixed the if

File: pkg/distributor/ha_tracker.go [hunk 4]
```diff
@@ -242,18 +243,21 @@ func replicasNotMatchError(replica, elected string) error {
...
-// the replica or cluster label and returning their values. Returns an error
-// if we find one but not both of the labels.
+// Searches for the values of replicaLabel and clusterLabel
+// returning their values. Returns an error if we find one but not both of the labels.
 func findHALabels(replicaLabel, clusterLabel string, labels []client.LabelAdapter) (string, string) {
 	var cluster, replica string
 	var pair client.LabelAdapter
 
-	for _, pair = range labels {
+	for i := 0; i < len(labels) && (replica == "" || cluster == ""); i++ {
```
Comment (pracucci) [inline pkg/distributor/ha_tracker.go]: The change set to this loop looks overcomplicated to me, just to break it earlier. I would suggest rolling back to the original implementation and simply switch `if pair.Name == clusterLabel {` to `if clusterLabel != "" && pair.Name == clusterLabel {`. I believe the performance you get breaking the loop earlier doesn't look justify making the loop way harder to read.

File: pkg/distributor/distributor.go [hunk 5]
```diff
@@ -299,6 +299,10 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
 
 	if d.limits.AcceptHASamples(userID) && len(req.Timeseries) > 0 {
 		cluster, replica := findHALabels(d.limits.HAReplicaLabel(userID), d.limits.HAClusterLabel(userID), req.Timeseries[0].Labels)
+		// If cluster has been explicitely set to "", it means we only care about the replica label, so use a placeholder value for KV
+		if cluster == "" {
```
Comment (pracucci) [inline pkg/distributor/distributor.go]: What if me move this logic to `findHALabels()`? Something like:

```
func findHALabels(replicaLabel, clusterLabel string, labels []client.LabelAdapter) (string, string) {
   // ...
   
   // <explain why we do it>
   if cluster == "" && clusterLabel == "" {
    cluster = fallbackClusterName
  }

   return cluster, replica
}
```

Does it make sense?

File: pkg/distributor/ha_tracker.go (hunk 2 above)
Comment (uepoch) [reply pkg/distributor/ha_tracker.go]: dne

File: pkg/distributor/distributor.go (hunk 5 above)
Comment (uepoch) [reply pkg/distributor/distributor.go]: I made a small change, it makes the loop cleaner, while keeping the breaking logic, does it looks ok ?

File: pkg/distributor/ha_tracker.go (hunk 4 above)
Comment (uepoch) [reply pkg/distributor/ha_tracker.go]: see below comment

**Comment by tomwilkie** [comment]
@uepoch do you mind adding the DCO two the last two commits?  We can rebase and deal with conflicts, but can't add the DCO.

**Comment by tomwilkie** [comment]
@uepoch do you mind adding the DCO two the last two commits?  We can rebase and deal with conflicts, but can't add the DCO.

**Comment by uepoch** [comment]
Done @tomwilkie 
Tried to fix what I could :)

**Comment by tomwilkie** [comment]
Thanks @uepoch! I think these changes LGTM, we just to "productionize" this change - would you mind:
- adding an entry to the changelog 
- updating the ha-pair-handing.md documentation to make it clear the cluster label is optional
- add a testcase to `TestDistributor_PushHAInstances` to show it working end-to-end

Thanks!

**Comment by uepoch** [comment]
Ok, will look at this tomorrow, sorry for the delay, i'm not currently working on cortex directly

**Comment by tomwilkie** [comment]
Thanks @uepoch - if you don't get a chance let me know and we can handle it.

**Comment by tomwilkie** [comment]
@uepoch I added the changelog, docs and tests, had to update your signed-off-by line to your new email; hope thats okay.

@jtlisi can you do final review please?

File: CHANGELOG.md [hunk 6]
```diff
@@ -84,6 +84,7 @@
 * [ENHANCEMENT] Experimental TSDB: added the flag `-experimental.tsdb.wal-compression-enabled` to allow to enable TSDB WAL compression. #2585
 * [ENHANCEMENT] Experimental TSDB: Querier and store-gateway components can now use so-called "caching bucket", which can currently cache fetched chunks into shared memcached server. #2572
 * [ENHANCEMENT] Experimental TSDB: added per-tenant blocks sharding support in the compactor, in order to parallelize blocks compaction for a single tenant in a single node. Sharding can be enabled via `-compactor.per-tenant-num-shards` while the parallelization can be controlled with `-compactor.per-tenant-shards-concurrency`. #2599
+* [ENHANCEMENT] Remove requirement for a cluster label value when deduping metrics. #1737
```
Comment (jtlisi) [inline CHANGELOG.md:87]: Can you note here that you have to set the `-distributor.ha-tracker.cluster` config value to an empty string to enable

**Comment by pstibrany** [comment]
Would you mind rebasing this, or do you want us to take over?

**Comment by pracucci** [comment]
_Going to close this PR because stale, but feel free to touch base if you're still interested in it._

**Pull request closed without merging**